File: java/src/main/java/cn/az/code/year2021/june/LowestCommonAncestorOfBinaryTree.java
package cn.az.code.year2021.june;

import cn.az.code.common.TreeNode;

/**
 * @author az
 * @since 2021-06-30
 */
public class LowestCommonAncestorOfBinaryTree {

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) {
            return root;
        }
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        return left == null ? right : right == null ? left : root;
    }
}

File: code/go/2021/june/lowest-common-ancestor-of-binary-tree.go
package june

import "leetcode/common"

func lowestCommonAncestor(root, p, q *common.TreeNode) *common.TreeNode {
	if root == nil || root == q || root == p {
		return root
	}
	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)
	if left != nil {
		if right != nil {
			return root
		}
		return left
	}
	return right
}
